<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <meta name="description" content="ConFindr Pipeline for detecting intra-species contamination in raw bacterial NGS reads.">
        
        <link rel="canonical" href="https://lowandrew.github.io/ConFindr/">
        <link rel="shortcut icon" href="./img/favicon.ico">

	<title>ConFindr</title>

        <link href="./css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="./css/font-awesome-4.0.3.css" rel="stylesheet">
        <link rel="stylesheet" href="./css/highlight.css">
        <link href="./css/base.css" rel="stylesheet">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <!-- Main title -->
            <a class="navbar-brand" href=".">ConFindr</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            <!-- Main navigation -->
            <ul class="nav navbar-nav">
            
            
                <li class="active">
                    <a href=".">Home</a>
                </li>
            
            
            
                <li >
                    <a href="install/">Installation</a>
                </li>
            
            
            
                <li >
                    <a href="usage/">Usage</a>
                </li>
            
            
            </ul>

            <!-- Search, Navigation and Repo links -->
            <ul class="nav navbar-nav navbar-right">
                <li class="disabled">
                    <a rel="next" >
                        <i class="fa fa-arrow-left"></i> Previous
                    </a>
                </li>
                <li >
                    <a rel="prev" href="install/">
                        Next <i class="fa fa-arrow-right"></i>
                    </a>
                </li>
                
                <li>
                    <a href="https://github.com/lowandrew/Confindr">
                        
                            <i class="fa fa-github"></i>
                        
                        GitHub
                    </a>
                </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#what-is-confindr">What is ConFindr?</a></li>
        
    
        <li class="main "><a href="#how-does-confindr-work">How Does ConFindr Work?</a></li>
        
            <li><a href="#brief-overview">Brief Overview</a></li>
        
            <li><a href="#detailed-overview-faq">Detailed Overview & FAQ</a></li>
        
    
    </ul>
</div></div>
            <div class="col-md-9" role="main">

<h1 id="what-is-confindr">What is ConFindr?</h1>
<p>ConFindr is a pipeline that can detect contamination in bacterial NGS data, both between and within species. It can do this with exceptional sensitivity - two samples mixed together with as few as 500 SNPs between them (&gt; 99.9 percent identity!) can be identified. This allows for stringent quality control of NGS samples.</p>
<h1 id="how-does-confindr-work">How Does ConFindr Work?</h1>
<h3 id="brief-overview">Brief Overview</h3>
<p>ADD IMAGE HERE EVENTUALLY</p>
<p>ConFindr works by looking at rMLST genes. These 53 genes are known to be single copy and conserved across all bacteria, making them excellent markers. As they are known to be single copy (with some caveats), any sample that has multiple alleles of one or more rMLST gene is likely to be contaminated. To identify the presence of multiple alleles in a sample, the following workflow is followed:</p>
<ol>
<li>Determine the genus of each sample so that genus-specific rMLST databases can be constructed.</li>
<li>Perform stringent quality trimming and bait out reads that contain rMLST gene sequence, using BBDuk.</li>
<li>Subsample rMLST reads to a depth of approximately 20X.</li>
<li>Split reads into kmers using jellyfish - typically to a size of k=31. </li>
<li>Compare all of the kmers found to all other kmers found, looking for pairs of kmers that differ only by one substitution - these are assumed to represent multiple alleles of the same gene.</li>
<li>Repeat steps 3-5 a few times (typically 5) and take the median number of contaminating kmers from these repetitions.</li>
<li>Use Mash to do a quick check for cross-species contamination.</li>
</ol>
<h3 id="detailed-overview-faq">Detailed Overview &amp; FAQ</h3>
<h5 id="why-subsample-reads">Why Subsample Reads?</h5>
<p>In our testing, the coverage depth of samples we tested varied wildly. Some samples had as little coverage as 20-30X, while others were well over 200X. We found that samples with very high coverage (over 100X) frequently had false positive results coming up, as some kmers were present often enough due to sequencing error to look like another allele was present. Enforcing a higher cutoff for number of kmers present on the high depth samples caused drastic cuts to sensitivity. Subsampling reads to 20X coverage prevents sequencing errors while maintaining good sensitivity.</p>
<h5 id="why-not-just-map-reads-to-the-rmlst-database">Why Not Just Map Reads to the rMLST Database?</h5>
<p>The obvious way to approach this problem would be to map reads to the rMLST database and look for hits to multiple alleles of the same gene. This approach was attempted in early iterations of the pipeline, but was found to work fairly poorly. The issue here was that many rMLST alleles are similar enough that the same read will map to many of them. Because of this, we would often only have one or two reads that mapped unambiguously to rMLST genes across a sample - not enough to reliably call contamination.</p></div>
        </div>

        <footer class="col-md-12">
            <hr>
            
            <center>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</center>
        </footer>

        <script src="./js/jquery-1.10.2.min.js"></script>
        <script src="./js/bootstrap-3.0.3.min.js"></script>
        <script src="./js/highlight.pack.js"></script>
        <script src="./js/base.js"></script>
    </body>
</html>