<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <meta name="description" content="ConFindr Pipeline for detecting intra-species contamination in raw bacterial NGS reads.">
        
        <link rel="canonical" href="https://lowandrew.github.io/ConFindr/">
        <link rel="shortcut icon" href="./img/favicon.ico">

	<title>ConFindr</title>

        <link href="./css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="./css/font-awesome-4.0.3.css" rel="stylesheet">
        <link rel="stylesheet" href="./css/highlight.css">
        <link href="./css/base.css" rel="stylesheet">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <!-- Main title -->
            <a class="navbar-brand" href=".">ConFindr</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            <!-- Main navigation -->
            <ul class="nav navbar-nav">
            
            
                <li class="active">
                    <a href=".">Home</a>
                </li>
            
            
            
                <li >
                    <a href="install/">Installation</a>
                </li>
            
            
            
                <li >
                    <a href="usage/">Usage</a>
                </li>
            
            
            </ul>

            <!-- Search, Navigation and Repo links -->
            <ul class="nav navbar-nav navbar-right">
                <li class="disabled">
                    <a rel="next" >
                        <i class="fa fa-arrow-left"></i> Previous
                    </a>
                </li>
                <li >
                    <a rel="prev" href="install/">
                        Next <i class="fa fa-arrow-right"></i>
                    </a>
                </li>
                
                <li>
                    <a href="https://github.com/lowandrew/Confindr">
                        
                            <i class="fa fa-github"></i>
                        
                        GitHub
                    </a>
                </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#what-is-confindr">What is ConFindr?</a></li>
        
            <li><a href="#how-does-confindr-work">How Does ConFindr Work?</a></li>
        
            <li><a href="#confindr-intra-species-contamination-detection-performance">ConFindr Intra-species Contamination Detection Performance</a></li>
        
            <li><a href="#confindr-inter-species-contamination-detection">ConFindr Inter-species Contamination Detection</a></li>
        
    
    </ul>
</div></div>
            <div class="col-md-9" role="main">

<h1 id="what-is-confindr">What is ConFindr?</h1>
<p>ConFindr is a pipeline that can detect contamination in bacterial NGS data, both between and within species. It can do this with exceptional sensitivity - two samples mixed together with as few as
500 SNPs between them (&gt; 99.9 percent identity!) can be identified. This allows for stringent quality control of NGS samples.</p>
<h2 id="how-does-confindr-work">How Does ConFindr Work?</h2>
<p>ConFindr works by looking at rMLST genes. These 53 genes are known to be single copy and conserved across all bacteria, making them excellent markers. As they are known to be single copy (with some caveats), any sample that has multiple alleles of one or more rMLST gene is likely to be contaminated. To identify the presence of multiple alleles in a sample, the following workflow is followed:</p>
<ol>
<li>Use Mash to determine the genus of each sample so that genus-specific rMLST databases can be constructed
and check for interspecies contamination.</li>
<li>Perform stringent quality trimming and bait out reads that contain rMLST gene sequence, using BBDuk.</li>
<li>Align reads back to the rMLST genes.</li>
<li>Look at the alignment to find <code>Contaminating SNVs</code> - those that have sites where more than one base is present, indicating
that multiple alleles are present, indicating contamination.</li>
</ol>
<h2 id="confindr-intra-species-contamination-detection-performance">ConFindr Intra-species Contamination Detection Performance</h2>
<p>The below graph shows the magnitude of contamination detected for several synthetic datasets in <em>Escherichia coli</em>.
Strains were mixed together that were either identical (and so should have no contamination), not identical but
have the same rMLST type (and so are contaminated, but beyond the limit of detection), two strains with the same serotype,
and therefore very closely related, or two strains of differing serotypes.</p>
<p>The black line on the graph represents our cutoff for contaminated samples - any sample with a magnitude above that
can reliably be called contaminated. As can be seen, two different serotypes from the same species are reliably detected
at contamination levels of 5 percent or higher, two strains of the same serotype are often detected at 10 percent
contamination and almost always at 20 percent contamination, and two strains that have the same rMLST or are identical
never have contamination detected. Results are very similar for other species.</p>
<p><img alt="alt text" src="./performance.png" title="ConFindr Performance" /></p>
<h2 id="confindr-inter-species-contamination-detection">ConFindr Inter-species Contamination Detection</h2>
<p>ConFindr seems to reliably be able to detect interspecies contamination at levels of 5 percent or above, but its checks
for interspecies contamination are not particularly rigorous. If you're very worried about interspecies contamination,
it would be a good idea to put your samples through some sort of metagenomics software (such as Kraken or Kaiju).</p></div>
        </div>

        <footer class="col-md-12">
            <hr>
            
            <center>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</center>
        </footer>

        <script src="./js/jquery-1.10.2.min.js"></script>
        <script src="./js/bootstrap-3.0.3.min.js"></script>
        <script src="./js/highlight.pack.js"></script>
        <script src="./js/base.js"></script>
    </body>
</html>