<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        
        <link rel="canonical" href="https://olc-bioinformatics.github.io/ConFindr/usage/">
        <link rel="shortcut icon" href="../img/favicon.ico">
        <title>Usage - ConFindr</title>
        <link href="../css/bootstrap.min.css" rel="stylesheet">
        <link href="../css/font-awesome.min.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/docco.min.css">

        <script src="../js/jquery-1.10.2.min.js" defer></script>
        <script src="../js/bootstrap.min.js" defer></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
        <script>hljs.initHighlightingOnLoad();</script> 
    </head>

    <body>
        <div class="navbar fixed-top navbar-expand-lg navbar-dark bg-primary">
            <div class="container">
                <a class="navbar-brand" href="..">ConFindr</a>
                <!-- Expander button -->
                <button type="button" class="navbar-toggler" data-toggle="collapse" data-target="#navbar-collapse">
                    <span class="navbar-toggler-icon"></span>
                </button>

                <!-- Expanded navigation -->
                <div id="navbar-collapse" class="navbar-collapse collapse">
                        <!-- Main navigation -->
                        <ul class="nav navbar-nav">
                            <li class="navitem">
                                <a href=".." class="nav-link">Home</a>
                            </li>
                            <li class="navitem">
                                <a href="../install/" class="nav-link">Installation</a>
                            </li>
                            <li class="navitem active">
                                <a href="./" class="nav-link">Usage</a>
                            </li>
                        </ul>

                    <ul class="nav navbar-nav ml-auto">
                        <li class="nav-item">
                            <a href="#" class="nav-link" data-toggle="modal" data-target="#mkdocs_search_modal">
                                <i class="fa fa-search"></i> Search
                            </a>
                        </li>
                            <li class="nav-item">
                                <a rel="prev" href="../install/" class="nav-link">
                                    <i class="fa fa-arrow-left"></i> Previous
                                </a>
                            </li>
                            <li class="nav-item">
                                <a rel="next" class="nav-link disabled">
                                    Next <i class="fa fa-arrow-right"></i>
                                </a>
                            </li>
                            <li class="nav-item">
                                <a href="https://github.com/OLC-Bioinformatics/Confindr/edit/master/docs/usage.md" class="nav-link"><i class="fa fa-github"></i> Edit on GitHub</a>
                            </li>
                    </ul>
                </div>
            </div>
        </div>

        <div class="container">
            <div class="row">
                    <div class="col-md-3"><div class="navbar-light navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
    <div class="navbar-header">
        <button type="button" class="navbar-toggler collapsed" data-toggle="collapse" data-target="#toc-collapse" title="Table of Contents">
            <span class="fa fa-angle-down"></span>
        </button>
    </div>

    
    <div id="toc-collapse" class="navbar-collapse collapse card bg-secondary">
        <ul class="nav flex-column">
            
            <li class="nav-item" data-level="2"><a href="#example-dataset" class="nav-link">Example Dataset</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            
            <li class="nav-item" data-level="2"><a href="#interpreting-confindr-results" class="nav-link">Interpreting ConFindr Results</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            
            <li class="nav-item" data-level="2"><a href="#using-confindr-in-a-python-script" class="nav-link">Using ConFindr in a Python Script</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            
            <li class="nav-item" data-level="2"><a href="#using-schemes-other-than-rmlst" class="nav-link">Using Schemes other than rMLST</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            
            <li class="nav-item" data-level="2"><a href="#optional-arguments" class="nav-link">Optional Arguments</a>
              <ul class="nav flex-column">
              </ul>
            </li>
        </ul>
    </div>
</div></div>
                    <div class="col-md-9" role="main">

<h2 id="example-dataset">Example Dataset</h2>
<p>An example dataset has been uploaded to FigShare. You can download it to your current working directory with the following command:</p>
<p><code>wget https://ndownloader.figshare.com/files/9972709 &amp;&amp; tar xf 9972709</code></p>
<p>This example dataset contains two different serotypes of <em>Escherichia coli</em> mixed together - it's about 80/20 split of O103 and O157. Contamination like this is difficult to detect
with regular tools - it's possible to pick up that it's two different strains, but it can be finicky. ConFindr, however, has no difficulty picking up the fact that this sample is contaminated.</p>
<p>In order to have ConFindr analyze this sample, the parameters you need to provide are:</p>
<ul>
<li><code>-i, --input_directory</code>: The path to a directory containing the reads, in FASTQ format, that you want analyzed. If you're using the example dataset, you'll want to enter <code>example-data</code></li>
<li><code>-o, --output_name</code>: The base name for your output. If you put <code>output</code> for this parameter, a folder called <code>output</code> will be created, and a file called <code>confindr_report.csv</code> with contamination
information will be created in this folder</li>
</ul>
<p>So, if the <code>example-data</code> folder were downloaded to your current working directory and you want to have an output folder called <code>output</code>, the command to run ConFindr would be:</p>
<p><code>confindr.py -i example-data -o output</code></p>
<p>You can use absolute or relative paths, and trailing slashes are also acceptable for the directories specified.
If ConFindr is properly installed, you should see something similar to the following appear on your terminal:</p>
<pre><code class="language-bash">  2019-04-02 15:06:04  Welcome to ConFindr 0.7.0! Beginning analysis of your samples... 
  2019-04-02 15:06:04  Could not find Escherichia_db_cgderived.fasta 
  2019-04-02 15:06:04  Could not find Listeria_db_cgderived.fasta 
  2019-04-02 15:06:04  Could not find Salmonella_db_cgderived.fasta 
  2019-04-02 15:06:04  Could not find refseq.msh 
  2019-04-02 15:06:04  Databases not present - downloading basic databases now... 
  2019-04-02 15:06:04  Downloading mash refseq sketch... 
  2019-04-02 15:06:07  Downloading cgMLST-derived data for Escherichia, Salmonella, and Listeria... 
  2019-04-02 15:06:12  Did not find rMLST databases, if you want to use ConFindr on genera other than Listeria, Salmonella, and Escherichia, you'll need to download them. Instructions are available at https://olc-bioinformatics.github.io/ConFindr/install/#downloading-confindr-databases

  2019-04-02 15:06:12  Beginning analysis of sample example... 
  2019-04-02 15:06:12  Checking for cross-species contamination... 
  2019-04-02 15:06:29  Extracting conserved core genes... 
  2019-04-02 15:06:37  Quality trimming... 
  2019-04-02 15:06:38  Detecting contamination... 
  2019-04-02 15:07:05  Done! Number of contaminating SNVs found: 214

  2019-04-02 15:07:05  Contamination detection complete! 
</code></pre>
<p>The run shouldn't take too long - depending on how powerful your machine is, it should be done in
one to two minutes (slightly longer if an <em>Escherichia</em> specific database has not yet been set up).
Once the run is done, you'll be able to inspect your results. Take a look at <code>output/confindr_report.csv</code>:
The <code>ContamStatus</code> column should read <code>True</code>, and the <code>NumContamSNVs</code> column should have a value of something close to 200.</p>
<p>In any future uses of ConFindr, databases will not need to be re-downloaded.</p>
<h2 id="interpreting-confindr-results">Interpreting ConFindr Results</h2>
<p>The results file that ConFindr produces is in comma-separated value (CSV) format, which can be opened by any spreadsheet application (Excel, LibreOffice, etc.) or your favorite text editor.</p>
<p>The file has the following headers: <code>Sample</code>, <code>Genus</code>, <code>NumContamSNVs</code>, <code>ContamStatus</code>, <code>PercentContam</code>, <code>PercentContamStandardDeviation</code>, and <code>BasesExamined</code>. Of these, ContamStatus is the most important - it will be <code>True</code> if a sample
is contaminated, and <code>False</code> if a sample is not contaminated. Detailed descriptions of each header follow.</p>
<ul>
<li><code>Sample</code>: The name of the sample. ConFindr will take everything before the first underscore (_) character to be the name of the sample, as done with samples coming from an Illumina MiSeq.</li>
<li><code>Genus</code>: The genus that ConFindr thinks your sample is. If ConFindr couldn't figure out what genus your sample is from, this will be NA.
If multiple genera were found, they will all be listed here, separated by a <code>:</code></li>
<li><code>NumContamSNVs</code>: The number of times ConFindr found sites with more than one base present.</li>
<li><code>ContamStatus</code>: The most important of all! Will read <code>True</code> if contamination is present in the sample, and <code>False</code> if contamination is not present. The result will be <code>True</code> if any of the following conditions are met:<ul>
<li>More than 1 contaminating SNV per 10000 base pairs examined was found.</li>
<li>There is cross contamination between genera.</li>
</ul>
</li>
<li><code>PercentContam</code>: Based on the depth of the minor variant for sites with multiple bases, ConFindr guesses
at what percent of your reads come from a contaminant. The more sequencing depth you have, the more accurate this will
get. For lower levels of contamination (around 5 percent) this tends to get overestimated, but the number gets more accurate as
contamination level increases, as well as sequencing depth.</li>
<li><code>PercentContamStandardDeviation</code>: The standard deviation of the percentage contamination estimate. Very high values may
indicate something strange is going on.</li>
<li><code>BasesExamined</code>: The number of bases ConFindr examined when making the contamination call. Will usally be around 20kb for rMLST databases,
 and will vary when other databases are used.</li>
<li><code>DatabaseDownloadDate</code>: Date that rMLST databases were downloaded, if you have them. As these are curated and updated regularly,
it's a good idea to re-run <code>confindr_database_setup</code> every now and then.</li>
</ul>
<p>ConFindr will also produce two CSV files for each sample - one called <code>samplename_contamination.csv</code>, which shows the contaminating
sites, and one called <code>samplename_rmlst.csv</code>, which shows ConFindr's guess at which allele is present for each rMLST gene.</p>
<h2 id="using-confindr-in-a-python-script">Using ConFindr in a Python Script</h2>
<p>In the event you'd rather integrate ConFindr into a script than run from the command line, here's how:</p>
<pre><code class="language-python">from confindr_src import confindr

# Find read files.
paired_reads = confindr.find_paired_reads('path_to_fastq_folder', forward_id='_R1', reverse_id='_R2')
# Run confindr. This assumes that you have already downloaded the databases. If you haven't,
# you can run confindr.check_for_databases_and_download(database_location='path/where/you/want/to/download, tmpdir='a/tmp/dir')
for pair in paired_reads:
    confindr.find_contamination(pair=pair,
                                forward_id='_R1', # change if yours is different
                                threads=4, 
                                output_folder='path/to/output',
                                databases_folder='path/to/databases')

</code></pre>
<h2 id="using-schemes-other-than-rmlst">Using Schemes other than rMLST</h2>
<p>As of ConFindr 0.4.4, ConFindr has the option to use a cgMLST scheme instead of an rMLST scheme for increased
contamination detection sensitivity. This hasn't been tested extensively, but looks to be working. Runtime is
increased by a factor of 2 or 3 compared to running against rMLST genes. </p>
<p>To use this option, you'll need a a cgMLST FASTA file - all FASTA headers should be in format &gt;genename_allele</p>
<p>In order to decrease computation time, clustering the cgMLST FASTA before running is recommended. CD-HIT on default
parameters does this fairly well.</p>
<p>cgMLST files that are already clustered are available for <em>Salmonella</em> and <em>Escherichia</em>. To get them:</p>
<p>Escherichia: <code>wget 'https://scist01.blob.core.windows.net/olc/Escherichia_cgmlst.fasta?sp=r&amp;st=2020-07-15T13:17:25Z&amp;se=2029-07-15T21:17:25Z&amp;spr=https&amp;sv=2019-10-10&amp;sr=b&amp;sig=zF01kDCgmsmBWJ0cSpyfYi6CLldIjalwU2RgsswKmmI%3D' -O Escherichia_cgmlst.fasta</code></p>
<p>Salmonella: <code>wget 'https://scist01.blob.core.windows.net/olc/Salmonella_cgmlst.fasta?sp=r&amp;st=2020-07-15T13:13:16Z&amp;se=2029-07-15T21:13:16Z&amp;spr=https&amp;sv=2019-10-10&amp;sr=b&amp;sig=ncG%2F5MzKt57p1BUdyFtnhJUk9Yfi6x3rFhSQWPlT2Ek%3D' -O Salmonella_cgmlst.fasta</code></p>
<p>When using a cgMLST database, ConFindr will use the provided scheme for all samples regardless of genus.</p>
<p>Actually calling ConFindr with cgMLST:</p>
<p><code>confindr.py -i folder-with-Escherichia-files -o cgmlst-output -cgmlst /path/to/Escherichia_cgmlst.fasta</code></p>
<h2 id="optional-arguments">Optional Arguments</h2>
<p>ConFindr has a few optional arguments that allow you to modify its other parameters. Optional arguments are:</p>
<ul>
<li><code>-t, --threads</code>: The number of threads to run ConFindr analysis with. The default is to use all threads available on your machine, and ConFindr scales very well with more threads, so it's recommended that this option be left at the default unless you need the computational resources for something else.</li>
<li><code>-d</code>, --databases`: Path to ConFindr databases. These will be downloaded automatically if not present.</li>
<li><code>-k</code>, --keep_files`: Set this flag to keep intermediate files. Useful if you want to do manual inspection of the BAM files
that ConFindr creates, which are deleted by default.</li>
<li><code>-fid, --forward_id</code>: The identifier for forward reads in your input FASTQ folder. By default, this is <code>_R1</code>. If you follow a different naming scheme, this is the parameter to change.</li>
<li><code>-rid, --reverse_id</code>: The identifier for reverse reads in your input FASTQ folder. By default, this is <code>_R2</code>. If you follow a different naming scheme, this is the parameter to change. </li>
<li><code>-v, --version</code>: Display ConFindr version and exit.</li>
<li><code>-verbosity, --verbosity</code>: How much you want printed to the screen. Choose <code>debug</code> to get some extra, or <code>warning</code> to
get almost nothing. Default is <code>info</code>.</li>
<li><code>-b</code>, <code>--base_cutoff</code>: The number of high-quality bases needed to call a site as multiallelic, and therefore 
contributing to contamination. Defaults to 2, which is usually sensitive without producing false positives.
If dealing with high depth samples, adding the <code>-bf</code> parameter set to around <code>0.05</code> is likely to be helpful in reducing
false positives.</li>
<li><code>-bf</code>, <code>--base_fraction_cutoff</code>: The proportion of high-quality bases needed to call a site as multiallelic, and therefore 
contributing to contamination. Must be between 0 and 1. Not used by default.</li>
<li><code>-q</code>, <code>--quality_cutoff</code>: The phred score a base needs to have before it's considered
trustworthy enough to contribute to a site being multiallelic. Defaults to 20, which should
be suitable for most purposes. </li>
<li><code>--rmlst</code>: By default, ConFindr will use custom core-gene derived datasets for <em>Escherichia</em>, <em>Listeria</em>, and <em>Salmonella</em>
instead of rMLST. Activate this flag to force use of rMLST genes for all genera.</li>
<li><code>--cross_details</code>: By default, when ConFindr finds cross-contaminated samples it stops analysis. Activate
this flag to have analysis of number of cSNVs continue in order to get an estimate of percentage contamination.</li>
</ul></div>
            </div>
        </div>

        <footer class="col-md-12">
            <hr>
            <p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>
        <script>
            var base_url = "..",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
        <script src="../js/base.js" defer></script>
        <script src="../search/main.js" defer></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
                <p>From here you can search these documents. Enter your search terms below.</p>
                <form>
                    <div class="form-group">
                        <input type="search" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results" data-no-results-text="No results found"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>

    </body>
</html>
