<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        
        <link rel="canonical" href="https://lowandrew.github.io/ConFindr/usage/">
        <link rel="shortcut icon" href="../img/favicon.ico">

	<title>Usage - ConFindr</title>

        <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="../css/font-awesome-4.0.3.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.css">
        <link href="../css/base.css" rel="stylesheet">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <!-- Main title -->
            <a class="navbar-brand" href="..">ConFindr</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            <!-- Main navigation -->
            <ul class="nav navbar-nav">
            
            
                <li >
                    <a href="..">Home</a>
                </li>
            
            
            
                <li >
                    <a href="../install/">Installation</a>
                </li>
            
            
            
                <li class="active">
                    <a href="./">Usage</a>
                </li>
            
            
            </ul>

            <!-- Search, Navigation and Repo links -->
            <ul class="nav navbar-nav navbar-right">
                <li >
                    <a rel="next" href="../install/">
                        <i class="fa fa-arrow-left"></i> Previous
                    </a>
                </li>
                <li class="disabled">
                    <a rel="prev" >
                        Next <i class="fa fa-arrow-right"></i>
                    </a>
                </li>
                
                <li>
                    <a href="https://github.com/lowandrew/Confindr">
                        
                            <i class="fa fa-github"></i>
                        
                        GitHub
                    </a>
                </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#example-dataset">Example Dataset</a></li>
        
    
        <li class="main "><a href="#interpreting-confindr-results">Interpreting ConFindr Results</a></li>
        
    
        <li class="main "><a href="#using-confindr-in-a-python-script">Using ConFindr in a Python Script</a></li>
        
    
        <li class="main "><a href="#using-schemes-other-than-rmlst">Using Schemes other than rMLST</a></li>
        
    
        <li class="main "><a href="#optional-arguments">Optional Arguments</a></li>
        
    
    </ul>
</div></div>
            <div class="col-md-9" role="main">

<h2 id="example-dataset">Example Dataset</h2>
<p>An example dataset has been uploaded to FigShare. You can download it to your current working directory with the following command:</p>
<p><code>wget https://ndownloader.figshare.com/files/9972709 &amp;&amp; tar xf 9972709</code></p>
<p>This example dataset contains two different serotypes of <em>Escherichia coli</em> mixed together - it's about 80/20 split of O103 and O157. Contamination like this is difficult to detect
with regular tools - it's possible to pick up that it's two different strains, but it can be finicky. ConFindr, however, has no difficulty picking up the fact that this sample is contaminated.</p>
<p>In order to have ConFindr analyze this sample, the parameters you need to provide are:</p>
<ul>
<li><code>-i, --input_directory</code>: The path to a directory containing the reads, in FASTQ format, that you want analyzed. If you're using the example dataset, you'll want to enter <code>example-data</code></li>
<li><code>-o, --output_name</code>: The base name for your output. If you put <code>output</code> for this parameter, a folder called <code>output</code> will be created, and a file called <code>confindr_report.csv</code> with contamination
information will be created in this folder</li>
</ul>
<p>So, if the <code>example-data</code> folder were downloaded to your current working directory and you want to have an output folder called <code>output</code>, the command to run ConFindr would be:</p>
<p><code>confindr.py -i example-data -o output</code></p>
<p>You can use absolute or relative paths, and trailing slashes are also acceptable for the directories specified.
If ConFindr is properly installed, you should see something similar to the following appear on your terminal:</p>
<pre><code class="bash">  2018-09-10 13:24:32  Welcome to ConFindr 0.4.3! Beginning analysis of your samples...
  2018-09-10 13:24:32  Databases not present. Downloading to /home/user/.confindr_db. This may take a few minutes...
  2018-09-10 13:35:46  Databases successfully downloaded...
  2018-09-10 13:35:46  Beginning analysis of sample example...
  2018-09-10 13:35:46  Checking for cross-species contamination...
  2018-09-10 13:35:59  Setting up genus-specific database for genus Escherichia...
  2018-09-10 13:36:58  Extracting rMLST genes...
  2018-09-10 13:37:02  Quality trimming...
  2018-09-10 13:37:02  Detecting contamination...
  2018-09-10 13:37:20  Done! Number of contaminating SNVs found: 24

  2018-09-10 13:37:20  Contamination detection complete!
</code></pre>

<p>The run shouldn't take too long - depending on how powerful your machine is, it should be done in
one to two minutes (slightly longer if an <em>Escherichia</em> specific database has not yet been set up).
Once the run is done, you'll be able to inspect your results. Take a look at <code>output/confindr_report.csv</code>:
The <code>ContamStatus</code> column should read <code>True</code>, and the <code>NumContamSNVs</code> column should have a value of something close to 15.</p>
<p>In any future uses of ConFindr, databases will not need to be re-downloaded.</p>
<h2 id="interpreting-confindr-results">Interpreting ConFindr Results</h2>
<p>The results file that ConFindr produces is in comma-separated value (CSV) format, which can be opened by any spreadsheet application (Excel, LibreOffice, etc.) or your favorite text editor.</p>
<p>The file has the following headers: <code>Sample</code>, <code>Genus</code>, <code>NumContamSNVs</code>, <code>ContamStatus</code>, <code>PercentContam</code>, <code>PercentContamStandardDeviation</code>, and <code>BasesExamined</code>. Of these, ContamStatus is the most important - it will be <code>True</code> if a sample
is contaminated, and <code>False</code> if a sample is not contaminated. Detailed descriptions of each header follow.</p>
<ul>
<li><code>Sample</code>: The name of the sample. ConFindr will take everything before the first underscore (_) character to be the name of the sample, as done with samples coming from an Illumina MiSeq.</li>
<li><code>Genus</code>: The genus that ConFindr thinks your sample is. If ConFindr couldn't figure out what genus your sample is from, this will be NA.
If multiple genera were found, they will all be listed here, separated by a <code>:</code></li>
<li><code>NumContamSNVs</code>: The number of times ConFindr found sites with more than one base present.</li>
<li><code>ContamStatus</code>: The most important of all! Will read <code>True</code> if contamination is present in the sample, and <code>False</code> if contamination is not present. The result will be <code>True</code> if any of the following conditions are met:<ul>
<li>3 or more contaminating SNVs are found. </li>
<li>There is cross contamination between genera.</li>
</ul>
</li>
<li><code>PercentContam</code>: Based on the depth of the minor variant for sites with multiple bases, ConFindr guesses
at what percent of your reads come from a contaminant. The more sequencing depth you have, the more accurate this will
get. In my experience, this will tend to be a slight overestimate, particularly for low depth samples.</li>
<li><code>PercentContamStandardDeviation</code>: The standard deviation of the percentage contamination estimate. Very high values may
indicate something strange is going on.</li>
<li><code>BasesExamined</code>: The number of bases ConFindr examined when making the contamination call. Will usally be around 20kb.</li>
</ul>
<p>ConFindr will also produce two CSV files for each sample - one called <code>samplename_contamination.csv</code>, which shows the contaminating
sites, and one called <code>samplename_rmlst.csv</code>, which shows ConFindr's guess at which allele is present for each rMLST gene.</p>
<h2 id="using-confindr-in-a-python-script">Using ConFindr in a Python Script</h2>
<p>In the event you'd rather integrate ConFindr into a script than run from the command line, here's how:</p>
<pre><code class="python">from confindr import confindr

# Find read files.
paired_reads = confindr.find_paired_reads('path_to_fastq_folder', forward_id='_R1', reverse_id='_R2')
# Run confindr. This assumes that you have already downloaded the databases. If you haven't,
# you can run confindr.check_for_databases_and_download(database_location='path/where/you/want/to/download, tmpdir='a/tmp/dir')
for pair in paired_reads:
    confindr.find_contamination(pair=pair,
                                forward_id='_R1', # change if yours is different
                                threads=4, 
                                output_folder='path/to/output',
                                databases_folder='path/to/databases')

</code></pre>

<h2 id="using-schemes-other-than-rmlst">Using Schemes other than rMLST</h2>
<p>As of ConFindr 0.4.4, ConFindr has the option to use a cgMLST scheme instead of an rMLST scheme for increased
contamination detection sensitivity. This hasn't been tested extensively, but looks to be working. Runtime is
increased by a factor of 2 or 3 compared to running against rMLST genes. </p>
<p>To use this option, you'll need a a cgMLST FASTA file - all FASTA headers should be in format &gt;genename_allele</p>
<p>In order to decrease computation time, clustering the cgMLST FASTA before running is recommended. CD-HIT on default
parameters does this fairly well.</p>
<p>cgMLST files that are already clustered are available for <em>Salmonella</em> and <em>Escherichia</em>. To get them:</p>
<p>Escherichia: <code>wget 'https://scist01.blob.core.windows.net/olc/Escherichia_cgmlst.fasta?sp=r&amp;st=2018-10-25T13:45:13Z&amp;se=2020-10-31T21:45:13Z&amp;spr=https&amp;sv=2017-11-09&amp;sig=0fgvcf6R%2BSSiz7gPm7KKm5M78wpjAPjGawhzt%2BlY7iE%3D&amp;sr=b' -O Escherichia_cgmlst.fasta</code></p>
<p>Salmonella: <code>wget 'https://scist01.blob.core.windows.net/olc/Salmonella_cgmlst.fasta?sp=r&amp;st=2018-10-25T13:35:21Z&amp;se=2020-10-31T21:35:21Z&amp;spr=https&amp;sv=2017-11-09&amp;sig=w5Pq9e4hsa6PGr458%2Bx4b5zqf4F2a6OUUL9H3ewQTNc%3D&amp;sr=b' -O Salmonella_cgmlst.fasta</code></p>
<p>When using a cgMLST database, ConFindr will use the provided scheme for all samples regardless of genus.</p>
<p>Actually calling ConFindr with cgMLST:</p>
<p><code>confindr.py -i folder-with-Escherichia-files -o cgmlst-output -cgmlst /path/to/Escherichia_cgmlst.fasta</code></p>
<h2 id="optional-arguments">Optional Arguments</h2>
<p>ConFindr has a few optional arguments that allow you to modify its other parameters. Optional arguments are:</p>
<ul>
<li><code>-t, --threads</code>: The number of threads to run ConFindr analysis with. The default is to use all threads available on your machine, and ConFindr scales very well with more threads, so it's recommended that this option be left at the default unless you need the computational resources for something else.</li>
<li><code>-d</code>, --databases`: Path to ConFindr databases. These will be downloaded automatically if not present.</li>
<li><code>-k</code>, --keep_files`: Set this flag to keep intermediate files. Useful if you want to do manual inspection of the BAM files
that ConFindr creates, which are deleted by default.</li>
<li><code>-fid, --forward_id</code>: The identifier for forward reads in your input FASTQ folder. By default, this is <code>_R1</code>. If you follow a different naming scheme, this is the parameter to change.</li>
<li><code>-rid, --reverse_id</code>: The identifier for reverse reads in your input FASTQ folder. By default, this is <code>_R2</code>. If you follow a different naming scheme, this is the parameter to change. </li>
<li><code>-v, --version</code>: Display ConFindr version and exit.</li>
<li><code>-verbosity, --verbosity</code>: How much you want printed to the screen. Choose <code>debug</code> to get some extra, or <code>warning</code> to
get almost nothing. Default is <code>info</code>.</li>
<li><code>-b</code>, <code>--base_cutoff</code>: The number of high-quality bases needed to call a site as multiallelic, and therefore 
contributing to contamination. Defaults to 2, which is usually sensitive without producing false positives.
If dealing with high depth samples, adding the <code>-bf</code> parameter set to around <code>0.05</code> is likely to be helpful in reducing
false positives.</li>
<li><code>-bf</code>, <code>--base_fraction_cutoff</code>: The proportion of high-quality bases needed to call a site as multiallelic, and therefore 
contributing to contamination. Must be between 0 and 1. Not used by default.</li>
<li><code>-q</code>, <code>--quality_cutoff</code>: The phred score a base needs to have before it's considered
trustworthy enough to contribute to a site being multiallelic. Defaults to 20, which should
be suitable for most purposes. </li>
</ul></div>
        </div>

        <footer class="col-md-12">
            <hr>
            
            <center>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</center>
        </footer>

        <script src="../js/jquery-1.10.2.min.js"></script>
        <script src="../js/bootstrap-3.0.3.min.js"></script>
        <script src="../js/highlight.pack.js"></script>
        <script src="../js/base.js"></script>
    </body>
</html>